modules/account-lib/src/coin/stx/transactionBuilder.ts

```diff
@@ -0,0 +1,251 @@
+    this._transaction.stxTransaction.setFee(new BigNum(this._fee.fee));
+    this._transaction.stxTransaction.setNonce(new BigNum(this._nonce));
+    for (const sig of this._signatures) {
+      await this.transaction.signWithSignatures(sig);
+    }
+    if (this._signatures.length === 0) {
+      if (this._numberSignatures > 1 && this._multiSignerKeyPairs.length !== this._numberSignatures + 1) {
+        throw new SigningError('Invalid number of signers for multi-sig');
+      }
+      await this.transaction.sign(this._multiSignerKeyPairs.slice(0, this._numberSignatures));
+      if (this._numberSignatures > 1) {
+        // multi-sig
+        await this.transaction.appendOrigin(this._multiSignerKeyPairs[this._numberSignatures]);
+      }
+    }
+
+    this._transaction.loadInputsAndOutputs();
+    return this._transaction;
+  }
+
+  /** @inheritdoc */
+  protected signImplementation(key: BaseKey): Transaction {
+    this.checkDuplicatedKeys(key);
+    const signer = new KeyPair({ prv: key.key });
+
+    // Signing the transaction is an operation that relies on all the data being set,
+    // so we set the source here and leave the actual signing for the build step
+    this._multiSignerKeyPairs.push(signer);
+    this._numberSignatures = this._multiSignerKeyPairs.length > 1 ? this._multiSignerKeyPairs.length - 1 : this._multiSignerKeyPairs.length;
+    return this.transaction;
+  }
+
+  /** @inheritdoc */
+  protected get transaction(): Transaction {
+    return this._transaction;
+  }
+
+  /** @inheritdoc */
+  protected set transaction(transaction: Transaction) {
+    this._transaction = transaction;
+  }
+
+  /**
+   * Validates that the given key is not already in this._multiSignerKeyPairs
+   *
+   * @param {BaseKey} key - The key to check
+   */
+  private checkDuplicatedKeys(key: BaseKey) {
+    this._multiSignerKeyPairs.forEach(_sourceKeyPair => {
+      if (_sourceKeyPair.getKeys().prv === key.key) {
+        throw new SigningError('Repeated sign: ' + key.key);
+      }
+    });
+  }
+
+  /**
+   * Set the transaction fees
+   *
+   * @param {BaseFee} fee The maximum gas to pay
+   * @returns {TransactionBuilder} This transaction builder
+   */
+  fee(fee: BaseFee): this {
+    this.validateValue(new BigNumber(fee.fee));
+    this._fee = fee;
+    return this;
+  }
+
+  network(stacksNetwork: StacksNetwork): this {
+    this._network = stacksNetwork;
+    return this;
+  }
+
+  nonce(n: number): this {
+    this._nonce = n;
+    return this;
+  }
+
+  /**
+   *  Set the memo
+   *
+   * @param {string} memo
+   * @returns {TransactionBuilder} This transaction builder
+   */
+  memo(memo: string): this {
+    this._memo = memo;
+    return this;
+  }
+
+  /**
+   *  Set the number of signatures for multi-sig
+   *
+   * @param {number} numSigns
+   * @returns {TransactionBuilder} This transaction builder
+   */
+  numberSignatures(numSigns: number): this {
+    this.validateValue(new BigNumber(numSigns));
```

Comment:
Seems odd that this works fine with `0`. Is this the right check here?

Reply:
Validating numSigns to be positive value.